### DISCUSSION TITLE
Docker Container Discussion

### DISCUSSION THREAD

**Discussion by slominskir** [389ds/389-ds-base#6058] [General]
Hi,
   Some thoughts on the Docker container "dscontainer":
   
   1. Would be nice if there was a "stop" command.   If you want to customize an image you can't easily stop the container after starting it.   Currently I'm resorting to [kill command with ps then sleep for a bit and hope](https://github.com/JeffersonLab/dirsrv/blob/ad4bc22d396c8e126462e1252962d7904e6a040f/scripts/once.sh#L29).  Seems like simply running dscontainer takes longer the first time so useful to pay that cost at image build time (`🎉 Instance setup complete` logs once).
   2.  Would be nice if there was support for initialing database contents.   Many "official" database Docker images and some defacto ones rely on a convention of a directory named `/docker-entrypoint-initdb.d` ([mariadb](https://hub.docker.com/_/mariadb), [mysql](https://hub.docker.com/_/mysql), [oracle (defacto)](https://hub.docker.com/r/gvenzl/oracle-xe), [postgres](https://hub.docker.com/_/postgres)).   See [Example dirsrv](https://github.com/JeffersonLab/dirsrv/blob/ad4bc22d396c8e126462e1252962d7904e6a040f/scripts/docker-entrypoint.sh#L33).
   3. Would be nice if there was a clean way to programmatically know when the container is ready for configuration.   It appears the `dscontainer -H` healthcheck indicates healthy before the dirsrv is really completely ready.  Well, unless you start dscontainer during the build to initialize it and get to the `🎉 Instance setup complete`.  After that it seems the healthcheck is then reliable.
   
   Thoughts?  Let me know if this is wrong place to discuss this (discussion 1!).
   
   Regards,
   
   Ryan

**Comment by vashirov** [comment]
Hi @slominskir,

this is the right place to discuss, and congrats on opening the first discussion here 🥇!

1. `dscontainer` handles SIGTERM signal to shutdown gracefully: https://github.com/389ds/389-ds-base/blob/f26ac0147f043fe130b974cb4c777df5b41f9e3a/src/lib389/cli/dscontainer#L346-L361 
Do you want something like `dscontainer stop` that would do the heavy lifting of finding the PID and sending the signal?
2. I agree, unfortunately we're not heavily investing in the DS container right now, so if you want to help us implement this, it would be greatly appreciated! In the meantime, you can check out my approach here https://github.com/389ds/ds-container/blob/2102e7e5c184f049ac839139dd9ee20bdf868181/fedora-sample_entries/init.sh#L26
3. IIRC, healthcheck is a bit racy and I have this workaround that works for me pretty well https://github.com/389ds/ds-container/blob/2102e7e5c184f049ac839139dd9ee20bdf868181/fedora-sample_entries/init.sh#L10-L24

**Reply by slominskir** [reply]
A stop function for dscontainer would be great.  It can be sneaky complicated with PID files, signals, and UNIX process hierarchies.  I started a draft pull for this as I was poking around the code, but it will take more effort to move it out of draft form: https://github.com/389ds/389-ds-base/pull/6062.   Feel free to make edits, comments, or take this over.

Sounds like you are open to a conventional init directory and enhancements to the healthcheck.  I'll investigate as time allows.

**Reply by slominskir** [reply]
After poking around a little more I noticed a few interesting things:

1. There appears to be a fairly large [existing Python API](https://github.com/389ds/389-ds-base/tree/main/src/lib389/lib389).   I haven't [located any Sphinx docs](https://github.com/389ds/389-ds-base/issues/1899) or other developer API docs yet though.  There may be a `stop()` method or equivalent already in the exiting lib.
2. The dscontainer app [sets a random password first](https://github.com/389ds/389-ds-base/blob/1f95b57fb8b03f72c7d2a2ce5b14eed5ba320407/src/lib389/cli/dscontainer#L226), then later sets the password to whatever was passed in by the user, and does so after the health check is already saying the server is ready (thus the race condition mentioned above).

**Reply by vashirov** [reply]
1. In a traditional installs we rely on the init system to stop the server. And existing method to stop the server in a container is [`kill_ds()`](https://github.com/389ds/389-ds-base/blob/1f95b57fb8b03f72c7d2a2ce5b14eed5ba320407/src/lib389/cli/dscontainer#L347), but it works only in the context of the existing `dscontainer` instance.
2. Healthcheck works over ldapi socket, so the password doesn't matter.

**Reply by slominskir** [reply]
I think it's a reasonable expectation that setting the Directory Manager password via environment variable would take effect before the healthcheck says the server is ready.    Ignoring the Directory Manager account and only using ldapi would avoid the race condition, but won't work for other containers (remote users) that want to interact (presumably waiting via depends_on "healthy" condition).  Further ldapi assumes the container is run as root, which as discussed below may be non-optimal.

**Reply by slominskir** [reply]
stop() function issue: https://github.com/389ds/389-ds-base/issues/6068

**Comment by slominskir** [comment]
As an aside, it seems PID 1 responsibilities (signal forwarding and zombie reaping) add complexity to Docker containers and adds a burden for all container developers to understand.   The dscontainer init probably shouldn't rely on the runtime engine init process for this since it is optional (opt-in) and non-standard, but at the same time, re-inventing the wheel in every Docker container project out there seems odd.  I think dscontainer may [already handle this](https://github.com/389ds/389-ds-base/blob/1f95b57fb8b03f72c7d2a2ce5b14eed5ba320407/src/lib389/cli/dscontainer#L62-L74), though it requires some effort to verify it is being done fully.  Docker run [--init](https://docs.docker.com/engine/reference/commandline/container_run/#init) and docker compose [init: true](https://docs.docker.com/compose/compose-file/05-services/#init) are the Docker runtime options.    Kubernetes doesn't appear to have a bare init option and it's a [controversial topic](https://github.com/kubernetes/kubernetes/issues/84210).  PID 1 responsibilities are often handled by systemd when running on Linux VM hosts, but in container it's historically been [tini](https://github.com/krallin/tini).

**Reply by slominskir** [reply]
Also worth mentioning dscontainer assumes [default of running as root](https://github.com/389ds/389-ds-base/blob/1f95b57fb8b03f72c7d2a2ce5b14eed5ba320407/docker/389-ds-fedora/Dockerfile).   Best practice is to use a [limited, non-root service account user](https://docs.docker.com/develop/develop-images/instructions/#user).   Not really an issue for me as I'm interested in dscontainer for development and test purposes only.  In production we use RHEL yum installed IdM (freeipa, powered by 389).

**Reply by vashirov** [reply]
DS is stateful application. `ns-slapd` (the main DS binary) doesn't handle the state inside the container, i.e. it can't initialize a new instance, if there is no data. Cloud native apps have this functionality inside the main binary, so they can be started as PID 1 and handle their own state. But DS is not cloud native. You have to have this complexity somewhere. So a wrapper `dscontainer` exists to handle that and forward signals from the container runtime.

> Also worth mentioning dscontainer assumes [default of running as root](https://github.com/389ds/389-ds-base/blob/1f95b57fb8b03f72c7d2a2ce5b14eed5ba320407/docker/389-ds-fedora/Dockerfile). 

This file is outdated and will be removed https://github.com/389ds/389-ds-base/issues/5733

> Best practice is to use a [limited, non-root service account user](https://docs.docker.com/develop/develop-images/instructions/#user). 

Modern container engines (Podman, CRI-O) leverage user namespaces, which provide much greater flexibility. See for example https://opensource.com/article/18/12/podman-and-user-namespaces and https://www.redhat.com/sysadmin/rootless-podman-user-namespace-modes.

**Reply by slominskir** [reply]
Remapping user namespace is great ([Docker does it too](https://docs.docker.com/engine/security/userns-remap/)), but just building your container image to not use root in the first place might be easier for end users.

I noticed the new [Dockerfile indirectly mentioned above](https://build.opensuse.org/package/view_file/home:firstyear/389-ds-container/Dockerfile?expand=1) installs client tools.   I'm extending 389ds/dirsrv:2.4 and [I must install client tools](https://github.com/JeffersonLab/dirsrv/blob/77f3ead9d19bb990ce0814b3f7960c8d8c7674eb/Dockerfile#L20) in my own image to get them.   How does the versioning work?  Is latest and most recent version (2.4) not matching?   Using latest tag is probably more trouble than it's worth (though it's a common convention).   Versioning doesn't appear to follow [semver](https://semver.org/).   Would be nice if versions pushed to DockerHub are fixed and don't change (create a new version on every change - I've noticed pull on fixed version results periodically in a new/different image).   Also looks like [quay.io](https://quay.io/repository/389ds/dirsrv) is starting to complicate things with alternate option that is completely unique.

**Reply by vashirov** [reply]
Please see https://github.com/389ds/389-ds-base/issues/5436 regarding container versioning. As I mentioned somewhere else, we're not investing in DS containerization at this time due to other priorities, so things are a bit unmaintained right now.

**Reply by slominskir** [reply]
Thanks.   Sounds like versioning/packaging has been discussed with good ideas, just nothing has changed yet.    Poking around I think I found the version that has client tools: it's probably: https://hub.docker.com/r/389ds/dirsrv_debug (as opposed to https://hub.docker.com/r/389ds/dirsrv)

**Comment by Firstyear** [comment]
> 1. pay that cost at image build time (`🎉 Instance setup complete` logs once).

You can't do this. We have to generate some internal id's and set a new directory manager password at startup, as well as being able to create user requested state. So this has to stay at "first run".

**Reply by slominskir** [reply]
It appears there is code in dscontainer that only runs once.  Perhaps due to this check: https://github.com/389ds/389-ds-base/blob/1f95b57fb8b03f72c7d2a2ce5b14eed5ba320407/src/lib389/cli/dscontainer#L223.    It's nice to pre-run this code.   I don't think I'm understanding your comment, what is it exactly we can't do because of internal IDs?   It's possible the official image could actually pre-run this during the original build even as far as I can tell.  Perhaps you are concerned with the case when a user wants to mount the /data directory to a volume?   In that case the mount would erase whatever is inside the container, true, and it would need to re-generate.  But in the case when a user does not mount the /data directory to a volume I don't see why it can't be pre-generated.

The Directory Manager password handling is odd.  I'm not sure we absolutely must be setting the Directory Manager password to random value first then setting it to user environment supplied value later on (though this only happens on first run, later runs it just overwrites).  I haven't dug into why it's doing that though.  It may make sense to reset it each run to allow users to pass the value in each time, but the timing of this with relation to the health check could be improved.

**Reply by Firstyear** [reply]
It's because a DM password has to be set iirc else the container won't start. So if we hardcoded this at build time we just introduced a remote root to anyone who "forgets" to reset or change this.

**Reply by progier389** [reply]
If it is the only reason we could have done things more simply by hardcoding an hash value that is impossible to match (by having characters that are not base64 like  nsslapd-rootpw: {PBKDF2-SHA512}*** INVALID *** INVALID *** INVALID *** INVALID
  ***) that provides a value but no one can use it to bind to the server. And these days we can use ldapi to reset rootpw

**Reply by slominskir** [reply]
A hard-coded Directory Manager password that satisfies the initialization requirement, but isn't useable (can't bind with it) sounds ideal.   Alternatively, or in addition, we could refuse to start the server unless a new Directory Manager password is passed in (not as flexible, but might be ok).  I've created an issue ticket to track the dscontainer Directory Manager handling in general, though specifically the race condition: https://github.com/389ds/389-ds-base/issues/6069
